### PR TITLE
Add `cupy.poly1d`

### PR DESCRIPTION
**Pull request by Dahlia-Chehata**

### CODE REVIEW

**Comment by Dahlia-Chehata** [comment]
I'll add tests later

File: cupy/lib/polynomial.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
+import re
+import warnings
+
+import cupy
+from cupy.core.overrides import set_module
+
+
+@set_module('cupy')
+class RankWarning(UserWarning):
```
Comment (asi1024) [inline cupy/lib/polynomial.py:9]: Use `numpy.RankWarning` instead.

File: cupy/lib/polynomial.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+            line1 = toadd1
+            line2 = toadd2
+        else:
+            line2 += partstr + ' ' * (len(power) - 1)
+            line1 += ' ' * (len(partstr) - 1) + power
+    output += line1 + '\n' + line2
+    return output + astr[n:]
+
+
+@set_module('cupy')
```
Comment (asi1024) [inline cupy/lib/polynomial.py:48]: Conventionally we don't override modules. For example, we don't override `cupy.core.core.ndarray`.

Comment (leofang) [reply cupy/lib/polynomial.py]: I think it's necessary for NumPy compatibility, see #3266. But it should be done in a separate PR.

File: cupy/lib/polynomial.pyx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+    @property
+    def variable(self):
+        return self._variable
+
+    @property
+    def order(self):
+        return self._coeffs.size - 1
+
+    # TODO
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:71]: We conventionally write TODOs like as the following format:
```# TODO(username): description`.```

File: cupy/lib/polynomial.pyx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    def __rdiv__(self, other):
+        if cupy.isscalar(other):
+            return poly1d(other / self.coeffs)
+
+    __rtruediv__ = __rdiv__
+
+    def __eq__(self, other):
+        if not isinstance(other, poly1d):
+            return NotImplemented
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:238]: Fix to raise `NotImplementedError`.

File: cupy/lib/polynomial.pyx [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                        thestr = "%s - %s" % (thestr, newstr[1:])
+                    else:
+                        thestr = "%s + %s" % (thestr, newstr)
+            else:
+                thestr = newstr
+        return _raise_power(thestr)
+
+    # TODO
+    def __call__(self, val):
+        pass
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:183]: Fix to raise NotImplementedError.

File: cupy/lib/polynomial.pyx [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+            raise ValueError('Polynomial must be 1d only.')
+        c_or_r = cupy.trim_zeros(c_or_r, trim='f')
+        if c_or_r.size == 0:
+            c_or_r = cupy.array([0.])
+        self._coeffs = c_or_r
+        if variable is None:
+            variable = 'x'
+        self._variable = variable
+
+    def __array__(self, t=None):
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:113]: We don't want to support `__array__` method to avoid implicit type conversion. Please fix to raise `TypeError` like `cupy.core.core.ndarray.__array__`.

File: cupy/lib/polynomial.pyx [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    def roots(self):
+        pass
+
+    @property
+    def _coeffs(self):
+        return self.__dict__['coeffs']
+
+    @_coeffs.setter
+    def _coeffs(self, coeffs):
+        self.__dict__['coeffs'] = coeffs
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:82]: Why are they needed?

File: cupy/lib/polynomial.pyx [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    def __init__(self, c_or_r, r=False, variable=None):
+        if isinstance(c_or_r, poly1d):
+            self._variable = c_or_r._variable
+            self._coeffs = c_or_r._coeffs
+
+            if set(c_or_r.__dict__) - set(self.__dict__):
+                msg = ('In the future extra properties will not be copied '
+                       'across when constructing one poly1d from another')
+                warnings.warn(msg, FutureWarning, stacklevel=2)
+                self.__dict__.update(c_or_r.__dict__)
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:97]: Why is this part needed?

File: cupy/lib/polynomial.pyx [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            line1 += ' ' * (len(partstr) - 1) + power
+    output += line1 + '\n' + line2
+    return output + astr[n:]
+
+
+class poly1d(object):
+    """A one-dimensional polynomial class.
+
+    Args:
+        c_or_r (cupy.ndarray or cupy.poly1d): The polynomial's
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:42]: There is no interface for conversion from `numpy.poly1d` to `cupy.poly1d` currently. How about supporting it here?

Comment (asi1024) [reply cupy/lib/polynomial.pyx:42]: I think it's OK to allow all `array_like` objects like `cupy.array` in this case.

File: cupy/lib/polynomial.pyx [hunk 10]
```diff
@@ -266,10 +278,61 @@ class poly1d(object):
...
+            if self.coeffs.dtype != out.coeffs.dtype:
+                raise TypeError(
+                    '{} poly1d cannot be obtained from {} poly1d'.format(
+                        out.coeffs.dtype, self.coeffs.dtype))
+            if self.coeffs.shape != out.coeffs.shape:
+                raise ValueError(
+                    'Shape mismatch. Expected shape: {}, '
+                    'actual shape: {}'.format(self.coeffs.shape,
+                                              out.coeffs.shape))
+            coeffs_cpu = out.coeffs
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:316]: The return value does not have the same id with `out`.
```suggestion
            out.coeffs = self.coeffs.get()
            return out
```

File: cupy/lib/polynomial.pyx [hunk 11]
```diff
@@ -266,10 +278,61 @@ class poly1d(object):
...
+            if stream is not None:
+                coeffs_gpu.data.copy_to_host_async(ptr,
+                                                   coeffs_gpu.nbytes, stream)
+            else:
+                stream_ptr = stream_module.get_current_stream_ptr()
+                if stream_ptr == 0:
+                    coeffs_gpu.data.copy_to_host(ptr, coeffs_gpu.nbytes)
+                else:
+                    coeffs_gpu.data.copy_to_host_async(ptr, coeffs_gpu.nbytes)
+        return numpy.poly1d(coeffs_cpu, variable=self.variable)
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:335]:
```suggestion
        else:
            return numpy.poly1d(self.coeffs.get(stream=stream), self.variable)
```

File: cupy/lib/polynomial.pyx (hunk 7 above)
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:87]: According to Scipy
>  our internal _coeffs property need to be backed by __dict__['coeffs'] for scipy to work correctly.

File: cupy/lib/polynomial.pyx (hunk 8 above)
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:102]: same

File: cupy/lib/polynomial.pyx [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+
+    def __repr__(self):
+        vals = repr(self.coeffs)
+        vals = vals[6:-1]
+        return "poly1d(%s)" % vals
+
+    def __len__(self):
+        return self.order
+
+    def __str__(self):
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:133]: How about returning `str(self.get())` simply?

File: cupy/lib/polynomial.pyx [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+        if variable is None:
+            variable = 'x'
+        self._variable = variable
+
+    def __array__(self, dtype=None):
+        raise TypeError(
+            'Implicit conversion to a NumPy array is not allowed. '
+            'Please use `.get()` to construct a NumPy array explicitly.')
+
+    def __repr__(self):
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:125]: How about returning `repr(self.get())` simply?

File: cupy/lib/polynomial.pyx (hunk 7 above)
Comment (asi1024) [reply cupy/lib/polynomial.pyx:87]: SciPy seems to have their original `scipy.poly1d` version, but it is just an alias of `numpy.poly1d` in recent versions. So we don't need this hack.

File: cupy/lib/polynomial.pyx [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+            else:
+                stream_ptr = stream_module.get_current_stream_ptr()
+                if stream_ptr == 0:
+                    coeffs_gpu.data.copy_to_host(ptr, coeffs_gpu.nbytes)
+                else:
+                    coeffs_gpu.data.copy_to_host_async(ptr, coeffs_gpu.nbytes)
+        return numpy.poly1d(coeffs_cpu, variable=self.variable)
+
+
+warnings.simplefilter('always', numpy.RankWarning)
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:338]: This line is not needed because `numpy.RankWarning` will be enabled to `always` when importing `numpy`.

File: cupy/lib/polynomial.pyx [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+            out (numpy.poly1d): Output object. In order to enable asynchronous
+                copy, the underlying memory should be a pinned memory.
+
+        Returns:
+            numpy.poly1d: Copy of poly1d object on host memory.
+
+        """
+        if out is not None:
+            if not isinstance(out, numpy.poly1d):
+                raise TypeError('Only numpy.poly1d can be obtained from'
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:305]:
```suggestion
                raise TypeError('Only numpy.poly1d can be obtained from '
```

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+import pytest
+import numpy
+
+import cupy
+from cupy import testing
+
+
+@testing.gpu
+class TestPoly1d(unittest.TestCase):
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:11]: Please add tests for each type of array_like inputs.

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_array_equal()
+    def test_poly1d_leading_zeros(self, xp, dtype):
+        a = xp.array([0, 0, 1, 2, 3], dtype)
+        return xp.poly1d(a).coeffs
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_equal()
+    def test_poly1d_variable(self, xp, dtype):
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:34]: Please add test also for `variable=None`.

File: cupy/lib/polynomial.pyx [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+        raise NotImplementedError
+
+    __rtruediv__ = __rdiv__
+
+    def __eq__(self, other):
+        if not isinstance(other, poly1d):
+            raise NotImplementedError
+        if self.coeffs.shape != other.coeffs.shape:
+            return False
+        return (self.coeffs == other.coeffs).all()
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:148]:
```suggestion
        return (self.coeffs == other.coeffs).all().get()
```

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_equal()
+    def test_poly1d_order(self, xp, dtype):
+        a = testing.shaped_arange((10,), xp, dtype)
+        return xp.poly1d(a).order
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_equal()
+    def test_poly1d_eq(self, xp, dtype):
+        a, b = self.make_poly1d(xp, dtype)
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:47]: make_poly1d generates only `a != b` arrays.

Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py:47]: Yes, both numpy and cupy will return False in this case

**Comment by Dahlia-Chehata** [comment]
@asi1024 I think I addressed all the comments ?

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 16 above)
Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py:11]: Please address this comment.

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 19 above)
Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py:60]: Please check also for `a == b` case.

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 16 above)
Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py:11]: I added tests for array and poly1d inputs. What other types of 'array_like' inputs should I test ?

Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py:11]: Please add tests for CPU to GPU conversion as I commented in https://github.com/cupy/cupy/pull/3466#discussion_r444669936.

Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py:11]: OK, I thought that `test_poly1d_get1` and `test_poly1d_get2` were enough.

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 19 above)
Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py]: Please address this comment also in __ne__ test.

File: cupy/lib/polynomial.pyx [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+    def coefficients(self):
+        return self.coeffs
+
+    @property
+    def o(self):
+        return self.order
+
+    def __init__(self, c_or_r, r=False, variable=None):
+        if isinstance(c_or_r, numpy.poly1d):
+            c_or_r._coeffs = cupy.asarray(c_or_r.coeffs)
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:71]: This line disrupts the input variable.
```py
>>> import numpy
>>> import cupy
>>> x = numpy.poly1d([1, 2, 3])
>>> x + numpy.poly1d([1, 2, 3])
poly1d([2, 4, 6])
>>> y = cupy.poly1d(x)
>>> x + numpy.poly1d([1, 2, 3])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  ...
TypeError: no implementation found for 'numpy.polyadd' on types that implement __array_function__: [<class 'cupy.core.core.ndarray'>, <class 'numpy.ndarray'>]
```

File: cupy/lib/polynomial.pyx [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+
+    @property
+    def coefficients(self):
+        return self.coeffs
+
+    @property
+    def o(self):
+        return self.order
+
+    def __init__(self, c_or_r, r=False, variable=None):
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:69]: Please implement for `r is True` case (or raise `NotImplementedError`).

File: cupy/lib/polynomial.pyx [hunk 22]
```diff
@@ -67,13 +67,17 @@ cdef class poly1d:
         return self.order
 
     def __init__(self, c_or_r, r=False, variable=None):
+        if isinstance(c_or_r, numpy.poly1d):
+           self._coeffs = cupy.empty(c_or_r.coeffs.shape,
+                                     c_or_r.coeffs.dtype)
+           self._coeffs.set(c_or_r.coeffs)
```
[the remark this answers is not in the document]
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx]: I think I modified that

File: cupy/lib/polynomial.pyx (hunk 20 above)
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:71]: array() ?

File: cupy/lib/polynomial.pyx [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+            if self.coeffs.dtype != out.coeffs.dtype:
+                raise TypeError(
+                    '{} poly1d cannot be obtained from {} poly1d'.format(
+                        out.coeffs.dtype, self.coeffs.dtype))
+            if self.coeffs.shape != out.coeffs.shape:
+                raise ValueError(
+                    'Shape mismatch. Expected shape: {}, '
+                    'actual shape: {}'.format(self.coeffs.shape,
+                                              out.coeffs.shape))
+            out.coeffs = self.coeffs.get()
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/polynomial.pyx:234]: And also `self.variable` should be updated.

File: cupy/lib/polynomial.pyx [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+            self._coeffs = cupy.asarray(c_or_r.coeffs)
+        if isinstance(c_or_r, poly1d):
+            self._coeffs = c_or_r._coeffs
+        if isinstance(c_or_r, (numpy.poly1d, poly1d)):
+            self._variable = c_or_r._variable
+            if variable is not None:
+                self._variable = variable
+            return
+        if isinstance(c_or_r, (numpy.ndarray, list)):
+            c_or_r = cupy.asarray(c_or_r)
```
[the remark this answers is not in the document]
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:80]: can I eliminate ` c_or_r = cupy.asarray(c_or_r)` and use `c_or_r = cupy.atleast_1d(c_or_r)` immediately ?

File: cupy/lib/polynomial.pyx [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+        # TODO(Dahlia-Chehata): if r: c_or_r = poly(c_or_r)
+        if r:
+            raise NotImplementedError
+        if c_or_r.ndim < 1:
+            c_or_r = cupy.atleast_1d(c_or_r)
+        if c_or_r.ndim > 1:
+            raise ValueError('Polynomial must be 1d only.')
```
[the remark this answers is not in the document]
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:87]: if https://github.com/cupy/cupy/pull/3466#discussion_r447678628 is applied, no need for this fix then

File: cupy/lib/polynomial.pyx (hunk 23 above)
[the remark this answers is not in the document]
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:234]: Why eliminating L222 - L224 ?

[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/polynomial.pyx:234]: Sorry, L222-224 should be left without eliminating.

File: cupy/lib/polynomial.pyx (hunk 24 above)
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/polynomial.pyx:80]: OK.

File: cupy/lib/polynomial.pyx [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+    # TODO(Dahlia-Chehata): use polymul for non-scalars
+    def __mul__(self, other):
+        if cupy.isscalar(other):
+            return poly1d(self.coeffs * other)
+        raise NotImplementedError
+
+    # TODO(Dahlia-Chehata): use polymul for non-scalars
+    def __rmul__(self, other):
+        if cupy.isscalar(other):
+            return poly1d(other * self.coeffs)
```
[the remark this answers is not in the document]
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:123]: How can I test rmul / rdiv ?

[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/polynomial.pyx:123]: I think `__rmul__` and `__rdiv__` are not needed. How about removing them?

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
@Dahlia-Chehata Could you check Jenkins failure?

**Comment by Dahlia-Chehata** [comment]
@asi1024 The tests pass locally. PTAL

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 27]
```diff
@@ -155,12 +155,14 @@ def test_poly1d_get2(self, dtype):
...
     @testing.for_all_dtypes()
     def test_poly1d_get3(self, dtype):
-        arr = testing.shaped_arange((5,), cupy, dtype)
+        a1 = testing.shaped_arange((5,), cupy, dtype)
+        b1 = cupy.poly1d(a1, variable='z')
         out = numpy.poly1d(numpy.empty((5,), dtype))
-        cupy.poly1d(arr, variable='z').get(out=out)
-        a = testing.shaped_arange((5,), numpy, dtype)
-        b = numpy.poly1d(a, variable='z')
-        assert out == b
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py]: Could you explain why the CI failed in the previous version?

Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py]: I think because of the missing `else`. When I printed out before the assert statement. I found that it is printed twice for each dtype, the second print was exactly the same as the wrong test result of Jenkins.
As for https://github.com/cupy/cupy/pull/3466/commits/55e82569b01f4c4dcc10a772cb0b075531861442, I don't think it has an effect.
Not sure though because tests pass locally in both cases.

File: cupy/lib/polynomial.pyx [hunk 28]
```diff
@@ -0,0 +1,251 @@
...
+        # TODO(Dahlia-Chehata): if r: c_or_r = poly(c_or_r)
+        if r:
+            raise NotImplementedError
+        if c_or_r.ndim < 1:
+            c_or_r = cupy.atleast_1d(c_or_r)
+        if c_or_r.ndim > 1:
+            raise ValueError('Polynomial must be 1d only.')
+        c_or_r = cupy.trim_zeros(c_or_r, trim='f')
+        if c_or_r.size == 0:
+            c_or_r = cupy.array([0.])
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/polynomial.pyx:84]: Did you address this comment?

[the remark this answers is not in the document]
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:84]: Yes

[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/polynomial.pyx:84]: Which test?

[the remark this answers is not in the document]
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:84]: Apparently I didn't address this properly, I added `test_poly1d_zero_dim` instead

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 27 above)
Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py]: If you think https://github.com/cupy/cupy/pull/3466/commits/72dc3e77eecb05feb092b6b5e3f321ae02acc650 fixed the issue, could you revert https://github.com/cupy/cupy/pull/3466/commits/55e82569b01f4c4dcc10a772cb0b075531861442?

**Comment by asi1024** [comment]
Jenkins, test this please.

File: tests/cupy_tests/lib_tests/test_polynomial.py (hunk 27 above)
Comment (Dahlia-Chehata) [reply tests/cupy_tests/lib_tests/test_polynomial.py]: If you think there is something with the test or `get()` itself, please point it out.

Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py]: The property method `coeffs` in the newest NumPy stable version (used in Jenkins test `cupy-py3`) returns `self._coeffs` itself:
https://github.com/numpy/numpy/blob/c54a1acb713497f8c763574a4ba35fdfc864e53f/numpy/lib/polynomial.py#L1110-L1113
However, that in NumPy 1.15.4 (used in `cupy-py35`) returns a copy, so `out.coeffs` is not updated.
https://github.com/numpy/numpy/blob/3430d78c01a3b9a19adad75f1acb5ae18286da73/numpy/lib/polynomial.py#L1036-L1039

Comment (asi1024) [reply tests/cupy_tests/lib_tests/test_polynomial.py]: How about dropping `out` keyword argument of `get` method?

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.
